### PR TITLE
filewatch: make the controller level-triggered rather than edge triggered

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas, @landism,

Please review the following commits I made in branch nicks/filewatch:

9060ee9702bcd9e7b938d452664bbbc4c6a98b64 (2022-01-05 13:21:21 -0500)
filewatch: make the controller level-triggered rather than edge triggered
This also ensures that we properly handle errors, rather than
crashing tilt

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

File: internal/controllers/core/cmd/controller_test.go [hunk 1]
```diff
@@ -924,21 +923,7 @@ func newFixture(t *testing.T) *fixture {
...
-			next, shutdown := q.Get()
-			if shutdown {
-				return
-			}
-
-			_, _ = c.Reconcile(f.Context(), next.(reconcile.Request))
-			q.Done(next)
-		}
-	}()
+	indexer.StartRequeuerForTesting(f.Context(), c.requeuer, c)
```
Comment (milas) [inline internal/controllers/core/cmd/controller_test.go:926]: Ooo nice!

File: internal/controllers/core/filewatch/controller.go [hunk 2]
```diff
@@ -211,40 +247,30 @@ func (c *Controller) dispatchFileChangesLoop(ctx context.Context, st store.RStor
...
 					watch.WindowsBufferSizeEnvVar,
 					watch.DesiredWindowsBufferSize(),
-					err)))
+					err))
 			} else if err.Error() == fsnotify.ErrEventOverflow.Error() {
-				st.Dispatch(store.NewErrorAction(fmt.Errorf("%s\nerror: %v", DetectedOverflowErrMsg, err)))
+				w.recordError(fmt.Errorf("%s\nerror: %v", DetectedOverflowErrMsg, err))
 			} else {
-				st.Dispatch(store.NewErrorAction(err))
+				w.recordError(err)
```
Comment (milas) [inline internal/controllers/core/filewatch/controller.go:262]: Hurray for no more fatal errors! 🙌

File: internal/controllers/core/filewatch/watcher.go [hunk 3]
```diff
@@ -42,15 +58,38 @@ func (w *watcher) cleanupWatch(ctx context.Context) {
 	if err := w.notify.Close(); err != nil {
 		logger.Get(ctx).Debugf("Failed to close notifier for %q: %v", w.name.String(), err)
 	}
+
+	w.restartBackoff = w.restartBackoff * 2
```
Comment (milas) [inline internal/controllers/core/filewatch/watcher.go:62]: Should we put a max on this?

File: internal/controllers/core/filewatch/watcher.go [hunk 4]
```diff
@@ -60,47 +99,6 @@ func (w *watcher) recordEvent(ctx context.Context, client ctrlclient.Client, st
...
-		// 	reconciliation and let that handle updating status
-		//
-		// historically, this method did a PUT/replace, which meant file events
-		// could be indefinitely delayed on optimistic concurrency failures due
-		// to lack of retry
-		//
-		// now, we PATCH assuming the spec matches at the time of fetch, but this
-		// means there is technically still a race if the spec changes between
-		// our fetch and server processing the update, where we'll post "stale"
-		// events, but this is less harmful than effectively missing files events
```
Comment (milas) [inline internal/controllers/core/filewatch/watcher.go:74]: ✅

**Review by milas — APPROVED** [review]

File: internal/controllers/core/filewatch/watcher.go (hunk 3 above)
Comment (nicks) [reply internal/controllers/core/filewatch/watcher.go:64]: sure, set a max to 5 minutes

**Pull request merged**